src/components/cuda_pw/cuda_common.c

```diff
@@ -0,0 +1,168 @@
+#include <stdio.h>
+#include <dlfcn.h>
+#include <papi.h>
+
+#include "cuda_common.h"
+
+int get_env_PAPI_CUDA_ROOT(void)
+{
+    PAPI_CUDA_ROOT_ENV = getenv("PAPI_CUDA_ROOT");
+    if (PAPI_CUDA_ROOT_ENV == NULL) {
+        return PAPI_ECMP;
+    }
+    return PAPI_OK;
+}
+
+int load_cuda_sym(void)
+{
+    int papiErr = PAPI_OK;
+    // char path_lib[PATH_MAX];
+    // int strErr;
+    // strErr = snprintf(path_lib, PATH_MAX-2, "%s/lib64/libcuda.so", PAPI_CUDA_ROOT);
+    // printf("path_lib: %s\n", path_lib);
+
+    // $PAPI_CUDA_ROOT/targets/x86_64-linux/lib/stubs/libcuda.so
+    dl1 = dlopen("libcuda.so", RTLD_NOW | RTLD_GLOBAL);
+    if (dl1 == NULL) {
+        fprintf(stderr, "ERROR: Loading libcuda.so failed.\n");
+        goto fn_fail;
+    }
+
+    cuCtxSetCurrentPtr = DLSYM_AND_CHECK(dl1, "cuCtxSetCurrent");
+    cuCtxGetCurrentPtr = DLSYM_AND_CHECK(dl1, "cuCtxGetCurrent");
+    cuCtxDestroyPtr = DLSYM_AND_CHECK(dl1, "cuCtxDestroy");
+    cuCtxCreatePtr = DLSYM_AND_CHECK(dl1, "cuCtxCreate");
+    cuCtxGetDevicePtr = DLSYM_AND_CHECK(dl1, "cuCtxGetDevice");
+    cuDeviceGetPtr = DLSYM_AND_CHECK(dl1, "cuDeviceGet");
+    cuDeviceGetCountPtr = DLSYM_AND_CHECK(dl1, "cuDeviceGetCount");
+    cuDeviceGetNamePtr = DLSYM_AND_CHECK(dl1, "cuDeviceGetName");
+    cuDevicePrimaryCtxRetainPtr = DLSYM_AND_CHECK(dl1, "cuDevicePrimaryCtxRetain");
+    cuDevicePrimaryCtxReleasePtr = DLSYM_AND_CHECK(dl1, "cuDevicePrimaryCtxRelease");
+    cuInitPtr = DLSYM_AND_CHECK(dl1, "cuInit");
+    cuGetErrorStringPtr = DLSYM_AND_CHECK(dl1, "cuGetErrorString");
+    cuCtxPopCurrentPtr = DLSYM_AND_CHECK(dl1, "cuCtxPopCurrent");
+    cuCtxPushCurrentPtr = DLSYM_AND_CHECK(dl1, "cuCtxPushCurrent");
+    cuCtxSynchronizePtr = DLSYM_AND_CHECK(dl1, "cuCtxSynchronize");
+    cuDeviceGetAttributePtr = DLSYM_AND_CHECK(dl1, "cuDeviceGetAttribute");
+
+    fn_exit:
+        return papiErr;
+    fn_fail:
+        papiErr = PAPI_ENOSUPP;
+        goto fn_exit;
+}
+
+int load_cudart_sym(void)
+{
+    int papiErr = PAPI_OK;
+    dl2 = dlopen("libcudart.so", RTLD_NOW | RTLD_GLOBAL);
+    if (dl2 == NULL) {
+        fprintf(stderr, "ERROR: Loading libcudart.so failed.\n");
+        goto fn_fail;
+    }
+
+    cudaGetDevicePtr = DLSYM_AND_CHECK(dl2, "cudaGetDevice");
+    cudaGetDeviceCountPtr = DLSYM_AND_CHECK(dl2, "cudaGetDeviceCount");
+    cudaGetDevicePropertiesPtr = DLSYM_AND_CHECK(dl2, "cudaGetDeviceProperties");
+    cudaDeviceGetAttributePtr = DLSYM_AND_CHECK(dl2, "cudaDeviceGetAttribute");
+    cudaSetDevicePtr = DLSYM_AND_CHECK(dl2, "cudaSetDevice");
+    cudaFreePtr = DLSYM_AND_CHECK(dl2, "cudaFree");
+    cudaDriverGetVersionPtr = DLSYM_AND_CHECK(dl2, "cudaDriverGetVersion");
+    cudaRuntimeGetVersionPtr = DLSYM_AND_CHECK(dl2, "cudaRuntimeGetVersion");
+
+    fn_exit:
+        return papiErr;
+    fn_fail:
+        papiErr = PAPI_ENOSUPP;
+        goto fn_exit;
+}
+
+int load_cupti_common_sym(void)
```

Comment:
Why this function deviates from the `load_<package>_sym` used above?

Reply:
this function is to load cupti symbols that are common to both perfworks and events APIs. There are separate functions in the respective APIs to load cupti perfworks symbols and cupti events symbols specifically.

Reply:
ok